R/run-nmtran.R

```diff
@@ -0,0 +1,330 @@
+      rlang::abort(
+        c(
+          glue("Found multiple default NONMEM versions ({nm_vers}) at `{config_path}`"),
+          "i" = "Please ensure only one version is set to the default"
+        )
+      )
+    }else if(length(default_nm) == 1){
+      default_nm <- default_nm[[1]]
+    }else{
+      # If no default, use the last one (likely higher version)
+      default_nm <- nm_config[[length(nm_config)]]
+    }
+  }
+
+  # Set NM-TRAN executable path
+  nm_path <- default_nm$home
+  nmtran_exe <- file.path(nm_path, "tr", "NMTRAN.exe")
+
+  # Check if nmtran_presort exists
+  presort_dir <- file.path(dirname(dirname(nmtran_exe)), "util")
+  nmtran_presort_exe <- file.path(presort_dir, "nmtran_presort")
+  run_presort <- unname(fs::file_exists(nmtran_presort_exe))
+
+  return(
+    list(
+      nmtran_exe = nmtran_exe,
+      nonmem_version = basename(default_nm$home),
+      cmd_args = cmd_args,
+      nmtran_presort_exe = if(run_presort) nmtran_presort_exe else NULL
+    )
+  )
+}
+
+#' Execute `NM-TRAN` in a given directory
+#'
+#' Execute `NM-TRAN` in a given directory. Also runs `nmtran_presort` if an
+#' executable is found.
+#'
+#' @param nmtran_exe Path to `NM-TRAN` executable.
+#' @param mod_path Path of a model to evaluate. Should be relative to `dir`.
+#' @param cmd_args A character vector of command line arguments for the `NM-TRAN`
+#'  execution call
+#' @param nmtran_presort_exe Path to `nmtran_presort` executable. Only available
+#' for `NONMEM` versions `nm74gf`, `nm74gf_nmfe`, and `nm75`. If provided, will run
+#' `nmtran_presort` before running `NM-TRAN`. Set to `NULL` to skip this step.
+#' @param dir Directory in which to execute the command.
+#'
+#' @keywords internal
+execute_nmtran <- function(
+    nmtran_exe,
+    mod_path,
+    cmd_args = c("0", "0", "2"),
+    nmtran_presort_exe = NULL,
+    dir = "."
+){
+  checkmate::assert_directory_exists(dir)
+  checkmate::assert_character(cmd_args, len = 3)
+
+  run_dir <- as.character(fs::path_real(dir))
+
+  # Check if nmtran_presort exists
+  run_presort <- !is.null(nmtran_presort_exe)
+
+  # Preprocess with nmtran_presort
+  if(run_presort){
+    presort.p <- processx::run(
+      command = nmtran_presort_exe, wd = run_dir, args = character(),
+      stdout = "|", stdin = file.path(run_dir, mod_path),
+      stderr_to_stdout = TRUE, error_on_status = FALSE
+    )
+
+    presort_status_val <- presort.p$status
+    if(is.na(presort_status_val)){
+      rlang::abort("nmtran_presort terminated unexpectedly")
+    }else if(presort_status_val != 0){
+      return(
+        list(
+          nmtran_model = file.path(run_dir, mod_path),
+          run_dir = run_dir,
+          status = "nmtran_presort failed. See errors.",
+          status_val = presort_status_val,
+          output_lines = presort.p$stdout
+        )
+      )
+    }
+
+    # Write the output to tempzzzz1 control stream
+    writeLines(presort.p$stdout, con = file.path(run_dir, "tempzzzz1.ctl"))
```

Comment:
`stderr_to_stdout = TRUE` is incorrect here. `nmtran_presort` takes the original control stream on stdin and writes the result to stdout for `NM-TRAN` to consume. `nmtree_presort` is free to write anything to stderr. If you send stderr to stdout, you risk corrupting the control stream.

I appreciate that you're trying to fit this `nmtran_presort` call into the same reporting approach as the main `NM-TRAN`, but it doesn't fit given the above. You could do some dancing to make it work, but I really don't think it's worth the effort.

This is especially true given that I don't think this call is likely to fail. Consider this: it doesn't fail when fed bbr's README.

```
$ /opt/NONMEM/nm75/util/nmtran_presort <README.Rmd | head -n2
---
output: github_document
$ /opt/NONMEM/nm75/util/nmtran_presort <README.Rmd >/dev/null
$ echo $?
0
```

Or even any file in the bbr repo!

```sh
#!/bin/sh

status=0
for f in $(git ls-files)
do
    /opt/NONMEM/nm75/util/nmtran_presort >/dev/null <"$f" || {
        status=$?
        printf >&2 'failed on %s\n' "$f"
    }
done
exit "$status"
```

```
$ ./presort-bbr-files
$ echo $?
0
```

So, I think you should just use `error_on_status = TRUE` here, and let `processx::run` bubble up any unexpected and rare failures.

Other comments:

-   no need to collect stdout and then write it to a file; you can just pass a file name as the `stdout` argument

-   I'd vote to not repeat nmfe's poor naming by using `tempzzzz1.ctl` as the file name, especially given the user will see this directory when they call `run_nmtran` with `clean = FALSE`.

All together I'm suggesting something like this:

<details>
<summary>patch</summary>

```diff
diff --git a/R/run-nmtran.R b/R/run-nmtran.R
index 0cfee6c4..df38166e 100644
--- a/R/run-nmtran.R
+++ b/R/run-nmtran.R
@@ -206,41 +206,24 @@ execute_nmtran <- function(
 
   run_dir <- as.character(fs::path_real(dir))
 
-  # Check if nmtran_presort exists
-  run_presort <- !is.null(nmtran_presort_exe)
-
-  # Preprocess with nmtran_presort
-  if(run_presort){
-    presort.p <- processx::run(
-      command = nmtran_presort_exe, wd = run_dir, args = character(),
-      stdout = "|", stdin = file.path(run_dir, mod_path),
-      stderr_to_stdout = TRUE, error_on_status = FALSE
+  nmtran_input <- file.path(run_dir, mod_path)
+  if (!is.null(nmtran_presort_exe)) {
+    nmtran_input <- file.path(
+      run_dir,
+      paste0(fs::path_ext_remove(mod_path), "_presort", fs::path_ext(mod_path))
+    )
+    processx::run(
+      nmtran_presort_exe,
+      stdin = file.path(run_dir, mod_path),
+      stdout = nmtran_input,
+      wd = run_dir,
+      error_on_status = TRUE
     )
-
-    presort_status_val <- presort.p$status
-    if(is.na(presort_status_val)){
-      rlang::abort("nmtran_presort terminated unexpectedly")
-    }else if(presort_status_val != 0){
-      return(
-        list(
-          nmtran_model = file.path(run_dir, mod_path),
-          run_dir = run_dir,
-          status = "nmtran_presort failed. See errors.",
-          status_val = presort_status_val,
-          output_lines = presort.p$stdout
-        )
-      )
-    }
-
-    # Write the output to tempzzzz1 control stream
-    writeLines(presort.p$stdout, con = file.path(run_dir, "tempzzzz1.ctl"))
   }
 
-  # Run NM-TRAN
-  stdin_file <- ifelse(run_presort, "tempzzzz1.ctl", mod_path)
   nmtran.p <- processx::run(
     command = nmtran_exe, wd = run_dir, args = cmd_args,
-    stdout = "|", stdin = file.path(run_dir, stdin_file),
+    stdout = "|", stdin = nmtran_input,
     stderr_to_stdout = TRUE, error_on_status = FALSE
   )
 
@@ -256,7 +239,7 @@ execute_nmtran <- function(
 
   # Tabulate NM-TRAN results
   nmtran_results <- list(
-    nmtran_model = stdin_file,
+    nmtran_model = basename(nmtran_input),
     run_dir = run_dir,
     status = status,
     status_val = status_val,
```

</details>

Reply:
Good points (dont disagree with anything) - made the suggested change